diff --git a/Cargo.toml b/Cargo.toml
index 3b1c0d6..0c735ba 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -31,4 +31,7 @@ shuttle = { version = "0.7.1", optional = true }
 thiserror = "2.0.9"
 
+serde = "1.0"
+test-fuzz = { path = "../../test-fuzz" }
+
 [target.'cfg(windows)'.dependencies]
 winapi = { version = "0.3", features = ["memoryapi", "sysinfoapi", "winnt", "errhandlingapi"], optional = true }
@@ -49,2 +52,4 @@ elf = "0.0.10"
 json = "0.12"
 test_utils = { path = "test_utils/" }
+
+[workspace]
diff --git a/src/memory_region.rs b/src/memory_region.rs
index de5ffe5..eeca621 100644
--- a/src/memory_region.rs
+++ b/src/memory_region.rs
@@ -34,5 +34,5 @@ use std::{
 
 /// The state of a memory region.
-#[derive(Debug, Copy, Clone, Default, Eq, PartialEq)]
+#[derive(Debug, Copy, Clone, Default, Eq, PartialEq, serde::Deserialize, serde::Serialize)]
 pub enum MemoryState {
     /// The memory region is readable
@@ -50,5 +50,5 @@ pub type MemoryCowCallback = Box<dyn Fn(u64) -> Result<u64, ()>>;
 
 /// Memory region for bounds checking and address translation
-#[derive(Default, Eq, PartialEq)]
+#[derive(Default, Eq, PartialEq, Clone)]
 #[repr(C, align(32))]
 pub struct MemoryRegion {
@@ -67,4 +67,85 @@ pub struct MemoryRegion {
 }
 
+#[derive(serde::Deserialize, serde::Serialize)]
+struct SerializableMemoryRegion {
+    pub host_mem_seed: Option<u64>,
+    pub vm_addr: u64,
+    pub vm_addr_end: u64,
+    pub len: u64,
+    pub vm_gap_shift: u8,
+    pub state: MemoryState,
+}
+
+impl From<SerializableMemoryRegion> for MemoryRegion {
+    fn from(value: SerializableMemoryRegion) -> Self {
+        use rand::{rngs::StdRng, RngCore, SeedableRng};
+        let SerializableMemoryRegion {
+            host_mem_seed,
+            vm_addr,
+            vm_addr_end,
+            len,
+            vm_gap_shift,
+            state,
+        } = value;
+        if len > 0x100000000 {
+            std::process::exit(0);
+        }
+        let mut bytes = vec![0u8; len as usize];
+        if let Some(seed) = host_mem_seed {
+            let mut rng: StdRng = SeedableRng::seed_from_u64(seed);
+            rng.fill_bytes(&mut bytes);
+        };
+        Self {
+            host_addr: Cell::new(bytes.leak().as_ptr() as u64),
+            vm_addr,
+            vm_addr_end,
+            len,
+            vm_gap_shift,
+            state: Cell::new(state),
+        }
+    }
+}
+
+impl From<&MemoryRegion> for SerializableMemoryRegion {
+    fn from(value: &MemoryRegion) -> Self {
+        let MemoryRegion {
+            host_addr: _,
+            vm_addr,
+            vm_addr_end,
+            len,
+            vm_gap_shift,
+            state,
+        } = value;
+        Self {
+            host_mem_seed: None,
+            vm_addr: *vm_addr,
+            vm_addr_end: *vm_addr_end,
+            len: *len,
+            vm_gap_shift: *vm_gap_shift,
+            state: state.get(),
+        }
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for MemoryRegion {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let value = SerializableMemoryRegion::deserialize(deserializer)?;
+        Ok(Self::from(value))
+    }
+}
+
+impl serde::Serialize for MemoryRegion {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let value = SerializableMemoryRegion::from(self);
+        value.serialize(serializer)
+    }
+}
+
 impl MemoryRegion {
     fn new(slice: &[u8], vm_addr: u64, vm_gap_size: u64, state: MemoryState) -> Self {
@@ -181,5 +262,5 @@ pub enum AccessType {
 
 /// Memory mapping based on eytzinger search.
-pub struct UnalignedMemoryMapping<'a> {
+pub struct UnalignedMemoryMapping {
     /// Mapped memory regions
     regions: Box<[MemoryRegion]>,
@@ -189,5 +270,5 @@ pub struct UnalignedMemoryMapping<'a> {
     cache: UnsafeCell<MappingCache>,
     /// VM configuration
-    config: &'a Config,
+    config: Config,
     /// Executable sbpf_version
     sbpf_version: SBPFVersion,
@@ -196,5 +277,86 @@ pub struct UnalignedMemoryMapping<'a> {
 }
 
-impl fmt::Debug for UnalignedMemoryMapping<'_> {
+#[derive(serde::Deserialize, serde::Serialize)]
+struct SerializableUnalignedMemoryMapping {
+    regions: Box<[MemoryRegion]>,
+    config: Config,
+    sbpf_version: SBPFVersion,
+}
+
+impl<'a> From<SerializableUnalignedMemoryMapping> for UnalignedMemoryMapping {
+    fn from(value: SerializableUnalignedMemoryMapping) -> Self {
+        let SerializableUnalignedMemoryMapping {
+            regions,
+            config,
+            sbpf_version,
+        } = value;
+        let region_addresses = regions
+            .iter()
+            .map(|region| region.vm_addr)
+            .collect::<Vec<_>>()
+            .into_boxed_slice();
+        Self {
+            regions,
+            region_addresses,
+            cache: UnsafeCell::new(MappingCache::new()),
+            config: config.clone(),
+            sbpf_version,
+            cow_cb: None,
+        }
+    }
+}
+
+impl<'a> From<&'a UnalignedMemoryMapping> for SerializableUnalignedMemoryMapping {
+    fn from(value: &'a UnalignedMemoryMapping) -> Self {
+        let UnalignedMemoryMapping {
+            regions,
+            region_addresses: _,
+            cache: _,
+            config,
+            sbpf_version,
+            cow_cb: _,
+        } = value;
+        Self {
+            regions: regions.clone(),
+            config: config.clone(),
+            sbpf_version: *sbpf_version,
+        }
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for UnalignedMemoryMapping {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let value = SerializableUnalignedMemoryMapping::deserialize(deserializer)?;
+        Ok(Self::from(value))
+    }
+}
+
+impl serde::Serialize for UnalignedMemoryMapping {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let value = SerializableUnalignedMemoryMapping::from(self);
+        value.serialize(serializer)
+    }
+}
+
+impl Clone for UnalignedMemoryMapping {
+    fn clone(&self) -> Self {
+        Self {
+            regions: self.regions.clone(),
+            region_addresses: self.region_addresses.clone(),
+            cache: UnsafeCell::new(unsafe { &*self.cache.get() as &MappingCache }.clone()),
+            config: self.config.clone(),
+            sbpf_version: self.sbpf_version,
+            cow_cb: None,
+        }
+    }
+}
+
+impl fmt::Debug for UnalignedMemoryMapping {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         f.debug_struct("UnalignedMemoryMapping")
@@ -215,5 +377,5 @@ impl fmt::Debug for UnalignedMemoryMapping<'_> {
 }
 
-impl<'a> UnalignedMemoryMapping<'a> {
+impl<'a> UnalignedMemoryMapping {
     fn construct_eytzinger_order(
         &mut self,
@@ -261,5 +423,5 @@ impl<'a> UnalignedMemoryMapping<'a> {
             region_addresses: vec![0; regions.len()].into_boxed_slice(),
             cache: UnsafeCell::new(MappingCache::new()),
-            config,
+            config: config.clone(),
             sbpf_version,
             cow_cb,
@@ -332,5 +494,5 @@ impl<'a> UnalignedMemoryMapping<'a> {
             None => {
                 return generate_access_violation(
-                    self.config,
+                    &self.config,
                     self.sbpf_version,
                     access_type,
@@ -347,5 +509,5 @@ impl<'a> UnalignedMemoryMapping<'a> {
         }
 
-        generate_access_violation(self.config, self.sbpf_version, access_type, vm_addr, len)
+        generate_access_violation(&self.config, self.sbpf_version, access_type, vm_addr, len)
     }
 
@@ -377,5 +539,5 @@ impl<'a> UnalignedMemoryMapping<'a> {
             None => {
                 return generate_access_violation(
-                    self.config,
+                    &self.config,
                     self.sbpf_version,
                     AccessType::Load,
@@ -420,5 +582,5 @@ impl<'a> UnalignedMemoryMapping<'a> {
 
         generate_access_violation(
-            self.config,
+            &self.config,
             self.sbpf_version,
             AccessType::Load,
@@ -457,5 +619,5 @@ impl<'a> UnalignedMemoryMapping<'a> {
             _ => {
                 return generate_access_violation(
-                    self.config,
+                    &self.config,
                     self.sbpf_version,
                     AccessType::Store,
@@ -499,5 +661,5 @@ impl<'a> UnalignedMemoryMapping<'a> {
 
         generate_access_violation(
-            self.config,
+            &self.config,
             self.sbpf_version,
             AccessType::Store,
@@ -526,5 +688,5 @@ impl<'a> UnalignedMemoryMapping<'a> {
         }
         Err(
-            generate_access_violation(self.config, self.sbpf_version, access_type, vm_addr, 0)
+            generate_access_violation(&self.config, self.sbpf_version, access_type, vm_addr, 0)
                 .unwrap_err(),
         )
@@ -549,15 +711,29 @@ impl<'a> UnalignedMemoryMapping<'a> {
 /// Memory mapping that uses the upper half of an address to identify the
 /// underlying memory region.
+#[derive(serde::Deserialize, serde::Serialize)]
 pub struct AlignedMemoryMapping<'a> {
     /// Mapped memory regions
     regions: Box<[MemoryRegion]>,
     /// VM configuration
+    #[serde(with = "test_fuzz::serde_ref")]
     config: &'a Config,
     /// Executable sbpf_version
     sbpf_version: SBPFVersion,
     /// CoW callback
+    #[serde(skip)]
     cow_cb: Option<MemoryCowCallback>,
 }
 
+impl Clone for AlignedMemoryMapping<'_> {
+    fn clone(&self) -> Self {
+        Self {
+            regions: self.regions.clone(),
+            config: self.config,
+            sbpf_version: self.sbpf_version,
+            cow_cb: None,
+        }
+    }
+}
+
 impl fmt::Debug for AlignedMemoryMapping<'_> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
@@ -728,5 +904,5 @@ impl<'a> AlignedMemoryMapping<'a> {
 
 /// Maps virtual memory to host memory.
-#[derive(Debug)]
+#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
 pub enum MemoryMapping<'a> {
     /// Used when address translation is disabled
@@ -736,5 +912,5 @@ pub enum MemoryMapping<'a> {
     Aligned(AlignedMemoryMapping<'a>),
     /// Memory mapping that allows mapping unaligned memory regions.
-    Unaligned(UnalignedMemoryMapping<'a>),
+    Unaligned(UnalignedMemoryMapping),
 }
 
@@ -906,5 +1082,5 @@ fn generate_access_violation(
 
 /// Fast, small linear cache used to speed up unaligned memory mapping.
-#[derive(Debug)]
+#[derive(Debug, Clone)]
 struct MappingCache {
     // The cached entries.
diff --git a/src/program.rs b/src/program.rs
index ce3b2ea..eff912c 100644
--- a/src/program.rs
+++ b/src/program.rs
@@ -10,5 +10,5 @@ use {
 
 /// Defines a set of sbpf_version of an executable
-#[derive(Debug, PartialEq, PartialOrd, Eq, Clone, Copy)]
+#[derive(Debug, PartialEq, PartialOrd, Eq, Clone, Copy, serde::Deserialize, serde::Serialize)]
 pub enum SBPFVersion {
     /// The legacy format
@@ -318,5 +318,5 @@ impl<C: ContextObject> std::fmt::Debug for BuiltinProgram<C> {
 #[macro_export]
 macro_rules! declare_builtin_function {
-    ($(#[$attr:meta])* $name:ident $(<$($generic_ident:tt : $generic_type:tt),+>)?, fn rust(
+    ($(#[$attr:meta])* $name:ident, fn rust $(<$($lifetime:tt),+>)? (
         $vm:ident : &mut $ContextObject:ty,
         $arg_a:ident : u64,
@@ -329,7 +329,9 @@ macro_rules! declare_builtin_function {
         $(#[$attr])*
         pub struct $name {}
+        #[test_fuzz::test_fuzz_impl]
         impl $name {
             /// Rust interface
-            pub fn rust $(<$($generic_ident : $generic_type),+>)? (
+            #[test_fuzz::test_fuzz(enable_in_production)]
+            pub fn rust $(<$($lifetime),+>)? (
                 $vm: &mut $ContextObject,
                 $arg_a: u64,
@@ -340,9 +342,12 @@ macro_rules! declare_builtin_function {
                 $memory_mapping: &mut $MemoryMapping,
             ) -> $Result {
+                if matches!($memory_mapping, MemoryMapping::Identity) {
+                    std::process::exit(0);
+                }
                 $($rust)*
             }
             /// VM interface
             #[allow(clippy::too_many_arguments)]
-            pub fn vm $(<$($generic_ident : $generic_type),+>)? (
+            pub fn vm $(<$($lifetime),+>)? (
                 $vm: *mut $crate::vm::EbpfVm<$ContextObject>,
                 $arg_a: u64,
@@ -354,5 +359,5 @@ macro_rules! declare_builtin_function {
                 use $crate::vm::ContextObject;
                 let vm = unsafe {
-                    &mut *($vm.cast::<u64>().offset(-($crate::vm::get_runtime_environment_key() as isize)).cast::<$crate::vm::EbpfVm<$ContextObject>>())
+                    &mut *(($vm as *mut u64).offset(-($crate::vm::get_runtime_environment_key() as isize)) as *mut $crate::vm::EbpfVm<$ContextObject>)
                 };
                 let config = vm.loader.get_config();
@@ -360,5 +365,5 @@ macro_rules! declare_builtin_function {
                     vm.context_object_pointer.consume(vm.previous_instruction_meter - vm.due_insn_count);
                 }
-                let converted_result: $crate::error::ProgramResult = Self::rust $(::<$($generic_ident),+>)?(
+                let converted_result: $crate::error::ProgramResult = Self::rust(
                     vm.context_object_pointer, $arg_a, $arg_b, $arg_c, $arg_d, $arg_e, &mut vm.memory_mapping,
                 ).map_err(|err| $crate::error::EbpfError::SyscallError(err)).into();
diff --git a/src/vm.rs b/src/vm.rs
index 98cedab..adef73f 100644
--- a/src/vm.rs
+++ b/src/vm.rs
@@ -49,5 +49,5 @@ pub fn get_runtime_environment_key() -> i32 {
 
 /// VM configuration settings
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
 pub struct Config {
     /// Maximum call depth
diff --git a/test_utils/Cargo.toml b/test_utils/Cargo.toml
index 50b7429..98ea054 100644
--- a/test_utils/Cargo.toml
+++ b/test_utils/Cargo.toml
@@ -9,2 +9,5 @@ publish = false
 libc = "0.2"
 solana-sbpf = { path = "../" }
+
+serde = "1.0"
+test-fuzz = { path = "../../../test-fuzz" }
diff --git a/test_utils/src/lib.rs b/test_utils/src/lib.rs
index c8ed12c..8b75d03 100644
--- a/test_utils/src/lib.rs
+++ b/test_utils/src/lib.rs
@@ -22,5 +22,5 @@ pub mod syscalls;
 
 /// Simple instruction meter for testing
-#[derive(Debug, Clone, Default)]
+#[derive(Debug, Clone, Default, serde::Deserialize, serde::Serialize)]
 pub struct TestContextObject {
     /// Contains the register state at every instruction in order of execution
diff --git a/test_utils/src/syscalls.rs b/test_utils/src/syscalls.rs
index 48c3a20..3ac7fe5 100644
--- a/test_utils/src/syscalls.rs
+++ b/test_utils/src/syscalls.rs
@@ -34,5 +34,5 @@ declare_builtin_function!(
     /// **unused**. Returns the number of bytes written.
     SyscallTracePrintf,
-    fn rust(
+    fn rust<'a>(
         _context_object: &mut TestContextObject,
         _arg1: u64,
@@ -41,5 +41,5 @@ declare_builtin_function!(
         arg4: u64,
         arg5: u64,
-        _memory_mapping: &mut MemoryMapping,
+        _memory_mapping: &mut MemoryMapping<'a>,
     ) -> Result<u64, Box<dyn std::error::Error>> {
         println!("bpf_trace_printf: {arg3:#x}, {arg4:#x}, {arg5:#x}");
@@ -62,5 +62,5 @@ declare_builtin_function!(
     /// each argument must be a `u64`.
     SyscallGatherBytes,
-    fn rust(
+    fn rust<'a>(
         _context_object: &mut TestContextObject,
         arg1: u64,
@@ -69,5 +69,5 @@ declare_builtin_function!(
         arg4: u64,
         arg5: u64,
-        _memory_mapping: &mut MemoryMapping,
+        _memory_mapping: &mut MemoryMapping<'a>,
     ) -> Result<u64, Box<dyn std::error::Error>> {
         Ok(arg1.wrapping_shl(32)
@@ -86,5 +86,5 @@ declare_builtin_function!(
     /// cases. Arguments 3 to 5 are unused.
     SyscallMemFrob,
-    fn rust(
+    fn rust<'a>(
         _context_object: &mut TestContextObject,
         vm_addr: u64,
@@ -93,5 +93,5 @@ declare_builtin_function!(
         _arg4: u64,
         _arg5: u64,
-        memory_mapping: &mut MemoryMapping,
+        memory_mapping: &mut MemoryMapping<'a>,
     ) -> Result<u64, Box<dyn std::error::Error>> {
         let host_addr: Result<u64, EbpfError> =
@@ -111,5 +111,5 @@ declare_builtin_function!(
     /// C-like `strcmp`, return 0 if the strings are equal, and a non-null value otherwise.
     SyscallStrCmp,
-    fn rust(
+    fn rust<'a>(
         _context_object: &mut TestContextObject,
         arg1: u64,
@@ -118,5 +118,5 @@ declare_builtin_function!(
         _arg4: u64,
         _arg5: u64,
-        memory_mapping: &mut MemoryMapping,
+        memory_mapping: &mut MemoryMapping<'a>,
     ) -> Result<u64, Box<dyn std::error::Error>> {
         // C-like strcmp, maybe shorter than converting the bytes to string and comparing?
@@ -149,5 +149,5 @@ declare_builtin_function!(
     /// Prints a NULL-terminated UTF-8 string.
     SyscallString,
-    fn rust(
+    fn rust<'a>(
         _context_object: &mut TestContextObject,
         vm_addr: u64,
@@ -156,5 +156,5 @@ declare_builtin_function!(
         _arg4: u64,
         _arg5: u64,
-        memory_mapping: &mut MemoryMapping,
+        memory_mapping: &mut MemoryMapping<'a>,
     ) -> Result<u64, Box<dyn std::error::Error>> {
         let host_addr: Result<u64, EbpfError> =
@@ -174,5 +174,5 @@ declare_builtin_function!(
     /// Prints the five arguments formated as u64 in decimal.
     SyscallU64,
-    fn rust(
+    fn rust<'a>(
         _context_object: &mut TestContextObject,
         arg1: u64,
@@ -181,5 +181,5 @@ declare_builtin_function!(
         arg4: u64,
         arg5: u64,
-        memory_mapping: &mut MemoryMapping,
+        memory_mapping: &mut MemoryMapping<'a>,
     ) -> Result<u64, Box<dyn std::error::Error>> {
         println!(
diff --git a/tests/execution.rs b/tests/execution.rs
index 7b69c1e..1e2c681 100644
--- a/tests/execution.rs
+++ b/tests/execution.rs
@@ -2537,5 +2537,5 @@ declare_builtin_function!(
     /// For test_nested_vm_syscall()
     SyscallNestedVm,
-    fn rust(
+    fn rust<'a>(
         _context_object: &mut TestContextObject,
         depth: u64,
@@ -2544,5 +2544,5 @@ declare_builtin_function!(
         _arg4: u64,
         _arg5: u64,
-        _memory_mapping: &mut MemoryMapping,
+        _memory_mapping: &mut MemoryMapping<'a>,
     ) -> Result<u64, Box<dyn std::error::Error>> {
         let (result, expected_result): (Result<u64, Box<dyn std::error::Error>>, ProgramResult) =
