diff --git a/Cargo.toml b/Cargo.toml
index 35f484a..e19b4f1 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -31,4 +31,7 @@ shuttle = { version = "0.7.1", optional = true }
 thiserror = "2.0.9"
 
+serde = "1.0"
+test-fuzz = { path = "../../test-fuzz" }
+
 [target.'cfg(windows)'.dependencies]
 winapi = { version = "0.3", features = ["memoryapi", "sysinfoapi", "winnt", "errhandlingapi"], optional = true }
@@ -49,2 +52,4 @@ elf = "0.0.10"
 json = "0.12"
 test_utils = { path = "test_utils/" }
+
+[workspace]
diff --git a/src/memory_region.rs b/src/memory_region.rs
index 5ca5cc5..83a6797 100644
--- a/src/memory_region.rs
+++ b/src/memory_region.rs
@@ -58,4 +58,85 @@ pub struct MemoryRegion {
 }

+#[derive(serde::Deserialize, serde::Serialize)]
+struct SerializableMemoryRegion {
+    host_mem_seed: Option<u64>,
+    vm_addr: u64,
+    len: u64,
+    vm_gap_shift: u8,
+    writable: bool,
+    access_violation_handler_payload: Option<u16>,
+}
+
+impl From<SerializableMemoryRegion> for MemoryRegion {
+    fn from(value: SerializableMemoryRegion) -> Self {
+        use rand::{rngs::StdRng, RngCore, SeedableRng};
+        let SerializableMemoryRegion {
+            host_mem_seed,
+            vm_addr,
+            len,
+            vm_gap_shift,
+            writable,
+            access_violation_handler_payload,
+        } = value;
+        if len > 0x100000000 {
+            std::process::exit(0);
+        }
+        let mut bytes = vec![0u8; len as usize];
+        if let Some(seed) = host_mem_seed {
+            let mut rng: StdRng = SeedableRng::seed_from_u64(seed);
+            rng.fill_bytes(&mut bytes);
+        };
+        Self {
+            host_addr: bytes.leak().as_ptr() as u64,
+            vm_addr,
+            len,
+            vm_gap_shift,
+            writable,
+            access_violation_handler_payload,
+        }
+    }
+}
+
+impl From<&MemoryRegion> for SerializableMemoryRegion {
+    fn from(value: &MemoryRegion) -> Self {
+        let MemoryRegion {
+            host_addr: _,
+            vm_addr,
+            len,
+            vm_gap_shift,
+            writable,
+            access_violation_handler_payload,
+        } = value;
+        Self {
+            host_mem_seed: None,
+            vm_addr: *vm_addr,
+            len: *len,
+            vm_gap_shift: *vm_gap_shift,
+            writable: *writable,
+            access_violation_handler_payload: access_violation_handler_payload.clone(),
+        }
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for MemoryRegion {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let value = SerializableMemoryRegion::deserialize(deserializer)?;
+        Ok(Self::from(value))
+    }
+}
+
+impl serde::Serialize for MemoryRegion {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let value = SerializableMemoryRegion::from(self);
+        value.serialize(serializer)
+    }
+}
+
 impl MemoryRegion {
     fn new(slice: &[u8], vm_addr: u64, vm_gap_size: u64, writable: bool) -> Self {
@@ -183,8 +264,10 @@ pub enum AccessType {
 
 /// Common parts of [UnalignedMemoryMapping] and [AlignedMemoryMapping]
+#[derive(serde::Deserialize, serde::Serialize)]
 pub struct CommonMemoryMapping {
     /// Mapped memory regions
     regions: Box<[MemoryRegion]>,
     /// Access violation handler
+    #[serde(skip, default = "boxed_default_access_violation_handler")]
     access_violation_handler: AccessViolationHandler,
     /// VM configuration
@@ -196,4 +279,21 @@ pub struct CommonMemoryMapping {
 }

+impl Clone for CommonMemoryMapping {
+    fn clone(&self) -> Self {
+        Self {
+            regions: self.regions.clone(),
+            access_violation_handler: boxed_default_access_violation_handler(),
+            allow_memory_region_zero: self.allow_memory_region_zero,
+            max_call_depth: self.max_call_depth,
+            stack_frame_size: self.stack_frame_size,
+            sbpf_version: self.sbpf_version,
+        }
+    }
+}
+
+fn boxed_default_access_violation_handler() -> AccessViolationHandler {
+    Box::new(default_access_violation_handler)
+}
+
 impl CommonMemoryMapping {
     fn generate_access_violation(
@@ -246,4 +346,39 @@ pub struct UnalignedMemoryMapping {
 }

+impl<'de> serde::Deserialize<'de> for UnalignedMemoryMapping {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let common = CommonMemoryMapping::deserialize(deserializer)?;
+        Ok(Self {
+            common,
+            region_addresses: Box::new([]),
+            region_index_lookup: Box::new([]),
+            cache: UnsafeCell::new(MappingCache::new()),
+        })
+    }
+}
+
+impl serde::Serialize for UnalignedMemoryMapping {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        self.common.serialize(serializer)
+    }
+}
+
+impl Clone for UnalignedMemoryMapping {
+    fn clone(&self) -> Self {
+        Self {
+            common: self.common.clone(),
+            region_addresses: self.region_addresses.clone(),
+            region_index_lookup: self.region_index_lookup.clone(),
+            cache: UnsafeCell::new(unsafe { &*self.cache.get() as &MappingCache }.clone()),
+        }
+    }
+}
+
 impl fmt::Debug for UnalignedMemoryMapping {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
@@ -356,4 +491,5 @@ impl<'a> UnalignedMemoryMapping {
 /// Memory mapping that uses the upper half of an address to identify the
 /// underlying memory region.
+#[derive(Clone, serde::Deserialize, serde::Serialize)]
 pub struct AlignedMemoryMapping {
     /// Common parts
@@ -460,5 +596,5 @@ impl AlignedMemoryMapping {
 
 /// Maps virtual memory to host memory.
-#[derive(Debug)]
+#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
 pub enum MemoryMapping {
     /// Used when address translation is disabled
@@ -646,5 +782,5 @@ impl MemoryMapping {
 
 /// Fast, small linear cache used to speed up unaligned memory mapping.
-#[derive(Debug)]
+#[derive(Debug, Clone)]
 struct MappingCache {
     // The cached entries.
diff --git a/src/program.rs b/src/program.rs
index af2642e..f5a69ec 100644
--- a/src/program.rs
+++ b/src/program.rs
@@ -11,5 +11,5 @@ use {
 
 /// Defines a set of sbpf_version of an executable
-#[derive(Debug, PartialEq, PartialOrd, Eq, Clone, Copy)]
+#[derive(Debug, PartialEq, PartialOrd, Eq, Clone, Copy, serde::Deserialize, serde::Serialize)]
 pub enum SBPFVersion {
     /// The legacy format
@@ -423,5 +423,5 @@ macro_rules! declare_builtin_function {
     (
         $(#[$attr:meta])*
-        $name:ident $(<$($generic_ident:tt : $generic_type:tt),+>)?,
+        $name:ident,
         fn rust(
             $vm:ident : &mut $ContextObject:ty,
@@ -440,4 +440,54 @@ macro_rules! declare_builtin_function {
             $($codegen:tt)*
         })?
+    ) => {
+        $(#[$attr])*
+        pub struct $name {}
+        #[test_fuzz::test_fuzz_impl]
+        impl
+            $crate::program::BuiltinFunctionDefinition<$ContextObject> for
+            $name
+        {
+            type Error = $Err;
+            #[test_fuzz::test_fuzz(enable_in_production)]
+            fn rust(
+                $vm: &mut $ContextObject,
+                $arg_a: u64,
+                $arg_b: u64,
+                $arg_c: u64,
+                $arg_d: u64,
+                $arg_e: u64,
+                $memory_mapping: &mut $MemoryMapping,
+            ) -> core::result::Result<$Ok, $Err> {
+                if matches!($memory_mapping, MemoryMapping::Identity) {
+                    std::process::exit(0);
+                }
+                $($rust)*
+            }
+            $(fn codegen(
+                $jit: &mut $crate::program::JitCompiler<$ContextObject2>,
+            ) {
+                $($codegen)*
+            })?
+        }
+    };
+    (
+        $(#[$attr:meta])*
+        $name:ident $(<$($generic_ident:tt : $generic_type:tt),+>)?,
+        fn rust(
+            $vm:ident : &mut $ContextObject:ty,
+            $arg_a:ident : u64,
+            $arg_b:ident : u64,
+            $arg_c:ident : u64,
+            $arg_d:ident : u64,
+            $arg_e:ident : u64,
+            $memory_mapping:ident : &mut $MemoryMapping:ty,
+        ) -> Result<$Ok:ty, $Err:ty> {
+            $($rust:tt)*
+        }
+        $(fn codegen(
+            $jit2:ident : &mut $crate::program::JitCompiler<$ContextObject3:ty>,
+        ) {
+            $($codegen2:tt)*
+        })?
     ) => {
         $(#[$attr])*
@@ -462,7 +512,7 @@ macro_rules! declare_builtin_function {
             }
             $(fn codegen(
-                $jit: &mut $crate::program::JitCompiler<$ContextObject2>,
+                $jit2: &mut $crate::program::JitCompiler<$ContextObject3>,
             ) {
-                $($codegen)*
+                $($codegen2)*
             })?
         }
diff --git a/src/vm.rs b/src/vm.rs
index 58d6b3e..b9bf426 100644
--- a/src/vm.rs
+++ b/src/vm.rs
@@ -59,5 +59,5 @@ pub enum ExecutionMode {
 
 /// VM configuration settings
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
 pub struct Config {
     /// Maximum call depth
diff --git a/test_utils/Cargo.toml b/test_utils/Cargo.toml
index ef2eed6..d29c1d7 100644
--- a/test_utils/Cargo.toml
+++ b/test_utils/Cargo.toml
@@ -9,2 +9,5 @@ publish = false
 libc = "0.2"
 solana-sbpf = { path = "../" }
+
+serde = "1.0"
+test-fuzz = { path = "../../../test-fuzz" }
diff --git a/test_utils/src/lib.rs b/test_utils/src/lib.rs
index 6f2a799..880ef16 100644
--- a/test_utils/src/lib.rs
+++ b/test_utils/src/lib.rs
@@ -23,5 +23,5 @@ pub mod syscalls;
 
 /// Simple instruction meter for testing
-#[derive(Debug, Clone, Default)]
+#[derive(Debug, Clone, Default, serde::Deserialize, serde::Serialize)]
 pub struct TestContextObject {
     /// Maximal amount of instructions which still can be executed
