diff --git a/Cargo.toml b/Cargo.toml
index f56084e..1fa2416 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -31,4 +31,7 @@ shuttle = { version = "0.7.1", optional = true }
 thiserror = "2.0.9"
 
+serde = "1.0"
+test-fuzz = { path = "../../test-fuzz" }
+
 [target.'cfg(windows)'.dependencies]
 winapi = { version = "0.3", features = ["memoryapi", "sysinfoapi", "winnt", "errhandlingapi"], optional = true }
@@ -49,2 +52,4 @@ elf = "0.0.10"
 json = "0.12"
 test_utils = { path = "test_utils/" }
+
+[workspace]
diff --git a/src/memory_region.rs b/src/memory_region.rs
index d71c9dc..b366405 100644
--- a/src/memory_region.rs
+++ b/src/memory_region.rs
@@ -58,4 +58,85 @@ pub struct MemoryRegion {
 }
 
+#[derive(serde::Deserialize, serde::Serialize)]
+struct SerializableMemoryRegion {
+    host_mem_seed: Option<u64>,
+    vm_addr: u64,
+    len: u64,
+    vm_gap_shift: u8,
+    writable: bool,
+    access_violation_handler_payload: Option<u16>,
+}
+
+impl From<SerializableMemoryRegion> for MemoryRegion {
+    fn from(value: SerializableMemoryRegion) -> Self {
+        use rand::{rngs::StdRng, RngCore, SeedableRng};
+        let SerializableMemoryRegion {
+            host_mem_seed,
+            vm_addr,
+            len,
+            vm_gap_shift,
+            writable,
+            access_violation_handler_payload,
+        } = value;
+        if len > 0x100000000 {
+            std::process::exit(0);
+        }
+        let mut bytes = vec![0u8; len as usize];
+        if let Some(seed) = host_mem_seed {
+            let mut rng: StdRng = SeedableRng::seed_from_u64(seed);
+            rng.fill_bytes(&mut bytes);
+        };
+        Self {
+            host_addr: bytes.leak().as_ptr() as u64,
+            vm_addr,
+            len,
+            vm_gap_shift,
+            writable,
+            access_violation_handler_payload,
+        }
+    }
+}
+
+impl From<&MemoryRegion> for SerializableMemoryRegion {
+    fn from(value: &MemoryRegion) -> Self {
+        let MemoryRegion {
+            host_addr: _,
+            vm_addr,
+            len,
+            vm_gap_shift,
+            writable,
+            access_violation_handler_payload,
+        } = value;
+        Self {
+            host_mem_seed: None,
+            vm_addr: *vm_addr,
+            len: *len,
+            vm_gap_shift: *vm_gap_shift,
+            writable: *writable,
+            access_violation_handler_payload: access_violation_handler_payload.clone(),
+        }
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for MemoryRegion {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let value = SerializableMemoryRegion::deserialize(deserializer)?;
+        Ok(Self::from(value))
+    }
+}
+
+impl serde::Serialize for MemoryRegion {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let value = SerializableMemoryRegion::from(self);
+        value.serialize(serializer)
+    }
+}
+
 impl MemoryRegion {
     fn new(slice: &[u8], vm_addr: u64, vm_gap_size: u64, writable: bool) -> Self {
@@ -173,16 +254,33 @@ pub enum AccessType {
 
 /// Common parts of [UnalignedMemoryMapping] and [AlignedMemoryMapping]
-pub struct CommonMemoryMapping<'a> {
+#[derive(serde::Deserialize, serde::Serialize)]
+pub struct CommonMemoryMapping {
     /// Mapped memory regions
     regions: Box<[MemoryRegion]>,
     /// Access violation handler
+    #[serde(skip, default = "boxed_default_access_violation_handler")]
     access_violation_handler: AccessViolationHandler,
     /// VM configuration
-    config: &'a Config,
+    config: Config,
     /// Executable sbpf_version
     sbpf_version: SBPFVersion,
 }
 
-impl CommonMemoryMapping<'_> {
+impl Clone for CommonMemoryMapping {
+    fn clone(&self) -> Self {
+        Self {
+            regions: self.regions.clone(),
+            access_violation_handler: boxed_default_access_violation_handler(),
+            config: self.config.clone(),
+            sbpf_version: self.sbpf_version,
+        }
+    }
+}
+
+fn boxed_default_access_violation_handler() -> AccessViolationHandler {
+    Box::new(default_access_violation_handler)
+}
+
+impl CommonMemoryMapping {
     fn generate_access_violation(
         &self,
@@ -223,7 +321,7 @@ impl CommonMemoryMapping<'_> {
 
 /// Memory mapping based on eytzinger search.
-pub struct UnalignedMemoryMapping<'a> {
+pub struct UnalignedMemoryMapping {
     /// Common parts
-    common: CommonMemoryMapping<'a>,
+    common: CommonMemoryMapping,
     /// Regions vm_addr fields in Eytzinger order
     region_addresses: Box<[u64]>,
@@ -234,5 +332,75 @@ pub struct UnalignedMemoryMapping<'a> {
 }
 
-impl fmt::Debug for UnalignedMemoryMapping<'_> {
+#[derive(serde::Deserialize, serde::Serialize)]
+struct SerializableUnalignedMemoryMapping {
+    common: CommonMemoryMapping,
+    region_addresses: Box<[u64]>,
+    region_index_lookup: Box<[usize]>,
+}
+
+impl From<SerializableUnalignedMemoryMapping> for UnalignedMemoryMapping {
+    fn from(value: SerializableUnalignedMemoryMapping) -> Self {
+        let SerializableUnalignedMemoryMapping {
+            common,
+            region_addresses,
+            region_index_lookup,
+        } = value;
+        Self {
+            common,
+            region_addresses,
+            region_index_lookup,
+            cache: UnsafeCell::new(MappingCache::new()),
+        }
+    }
+}
+
+impl<'a> From<&'a UnalignedMemoryMapping> for SerializableUnalignedMemoryMapping {
+    fn from(value: &'a UnalignedMemoryMapping) -> Self {
+        let UnalignedMemoryMapping {
+            common,
+            region_addresses,
+            region_index_lookup,
+            cache: _,
+        } = value;
+        Self {
+            common: common.clone(),
+            region_addresses: region_addresses.clone(),
+            region_index_lookup: region_index_lookup.clone(),
+        }
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for UnalignedMemoryMapping {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let value = SerializableUnalignedMemoryMapping::deserialize(deserializer)?;
+        Ok(Self::from(value))
+    }
+}
+
+impl serde::Serialize for UnalignedMemoryMapping {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let value = SerializableUnalignedMemoryMapping::from(self);
+        value.serialize(serializer)
+    }
+}
+
+impl Clone for UnalignedMemoryMapping {
+    fn clone(&self) -> Self {
+        Self {
+            common: self.common.clone(),
+            region_addresses: self.region_addresses.clone(),
+            region_index_lookup: self.region_index_lookup.clone(),
+            cache: UnsafeCell::new(unsafe { &*self.cache.get() as &MappingCache }.clone()),
+        }
+    }
+}
+
+impl fmt::Debug for UnalignedMemoryMapping {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         f.debug_struct("UnalignedMemoryMapping")
@@ -243,5 +411,5 @@ impl fmt::Debug for UnalignedMemoryMapping<'_> {
 }
 
-impl<'a> UnalignedMemoryMapping<'a> {
+impl<'a> UnalignedMemoryMapping {
     fn construct_eytzinger_order(&mut self, mut in_index: usize, out_index: usize) -> usize {
         if out_index >= self.common.regions.len() {
@@ -282,5 +450,5 @@ impl<'a> UnalignedMemoryMapping<'a> {
                 regions: regions.into_boxed_slice(),
                 access_violation_handler,
-                config,
+                config: config.clone(),
                 sbpf_version,
             },
@@ -340,10 +508,11 @@ impl<'a> UnalignedMemoryMapping<'a> {
 /// Memory mapping that uses the upper half of an address to identify the
 /// underlying memory region.
-pub struct AlignedMemoryMapping<'a> {
+#[derive(Clone, serde::Deserialize, serde::Serialize)]
+pub struct AlignedMemoryMapping {
     /// Common parts
-    common: CommonMemoryMapping<'a>,
+    common: CommonMemoryMapping,
 }
 
-impl fmt::Debug for AlignedMemoryMapping<'_> {
+impl fmt::Debug for AlignedMemoryMapping {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         f.debug_struct("AlignedMemoryMapping")
@@ -353,5 +522,5 @@ impl fmt::Debug for AlignedMemoryMapping<'_> {
 }
 
-impl<'a> AlignedMemoryMapping<'a> {
+impl<'a> AlignedMemoryMapping {
     /// Creates a new MemoryMapping structure from the given regions
     pub fn new_with_access_violation_handler(
@@ -377,5 +546,5 @@ impl<'a> AlignedMemoryMapping<'a> {
                 regions: regions.into_boxed_slice(),
                 access_violation_handler,
-                config,
+                config: config.clone(),
                 sbpf_version,
             },
@@ -415,16 +584,16 @@ impl<'a> AlignedMemoryMapping<'a> {
 
 /// Maps virtual memory to host memory.
-#[derive(Debug)]
-pub enum MemoryMapping<'a> {
+#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
+pub enum MemoryMapping {
     /// Used when address translation is disabled
     Identity,
     /// Aligned memory mapping which uses the upper half of an address to
     /// identify the underlying memory region.
-    Aligned(AlignedMemoryMapping<'a>),
+    Aligned(AlignedMemoryMapping),
     /// Memory mapping that allows mapping unaligned memory regions.
-    Unaligned(UnalignedMemoryMapping<'a>),
+    Unaligned(UnalignedMemoryMapping),
 }
 
-impl<'a> MemoryMapping<'a> {
+impl<'a> MemoryMapping {
     pub(crate) fn new_identity() -> Self {
         MemoryMapping::Identity
@@ -599,5 +768,5 @@ impl<'a> MemoryMapping<'a> {
 
 /// Fast, small linear cache used to speed up unaligned memory mapping.
-#[derive(Debug)]
+#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
 struct MappingCache {
     // The cached entries.
diff --git a/src/program.rs b/src/program.rs
index 79705c4..70ca947 100644
--- a/src/program.rs
+++ b/src/program.rs
@@ -10,5 +10,5 @@ use {
 
 /// Defines a set of sbpf_version of an executable
-#[derive(Debug, PartialEq, PartialOrd, Eq, Clone, Copy)]
+#[derive(Debug, PartialEq, PartialOrd, Eq, Clone, Copy, serde::Deserialize, serde::Serialize)]
 pub enum SBPFVersion {
     /// The legacy format
@@ -323,5 +323,5 @@ impl<C: ContextObject> std::fmt::Debug for BuiltinProgram<C> {
 #[macro_export]
 macro_rules! declare_builtin_function {
-    ($(#[$attr:meta])* $name:ident $(<$($generic_ident:tt : $generic_type:tt),+>)?, fn rust(
+    ($(#[$attr:meta])* $name:ident, fn rust $(<$($lifetime:tt),+>)? (
         $vm:ident : &mut $ContextObject:ty,
         $arg_a:ident : u64,
@@ -334,7 +334,9 @@ macro_rules! declare_builtin_function {
         $(#[$attr])*
         pub struct $name {}
+        #[test_fuzz::test_fuzz_impl]
         impl $name {
             /// Rust interface
-            pub fn rust $(<$($generic_ident : $generic_type),+>)? (
+            #[test_fuzz::test_fuzz(enable_in_production)]
+            pub fn rust $(<$($lifetime),+>)? (
                 $vm: &mut $ContextObject,
                 $arg_a: u64,
@@ -345,9 +347,12 @@ macro_rules! declare_builtin_function {
                 $memory_mapping: &mut $MemoryMapping,
             ) -> $Result {
+                if matches!($memory_mapping, MemoryMapping::Identity) {
+                    std::process::exit(0);
+                }
                 $($rust)*
             }
             /// VM interface
             #[allow(clippy::too_many_arguments)]
-            pub fn vm $(<$($generic_ident : $generic_type),+>)? (
+            pub fn vm $(<$($lifetime),+>)? (
                 $vm: *mut $crate::vm::EbpfVm<$ContextObject>,
                 $arg_a: u64,
@@ -359,5 +364,5 @@ macro_rules! declare_builtin_function {
                 use $crate::vm::ContextObject;
                 let vm = unsafe {
-                    &mut *($vm.cast::<u64>().offset(-($crate::vm::get_runtime_environment_key() as isize)).cast::<$crate::vm::EbpfVm<$ContextObject>>())
+                    &mut *(($vm as *mut u64).offset(-($crate::vm::get_runtime_environment_key() as isize)) as *mut $crate::vm::EbpfVm<$ContextObject>)
                 };
                 let config = vm.loader.get_config();
@@ -365,5 +370,5 @@ macro_rules! declare_builtin_function {
                     vm.context_object_pointer.consume(vm.previous_instruction_meter - vm.due_insn_count);
                 }
-                let converted_result: $crate::error::ProgramResult = Self::rust $(::<$($generic_ident),+>)?(
+                let converted_result: $crate::error::ProgramResult = Self::rust(
                     vm.context_object_pointer, $arg_a, $arg_b, $arg_c, $arg_d, $arg_e, &mut vm.memory_mapping,
                 ).map_err(|err| $crate::error::EbpfError::SyscallError(err)).into();
diff --git a/src/vm.rs b/src/vm.rs
index cb068ca..5c6d7f3 100644
--- a/src/vm.rs
+++ b/src/vm.rs
@@ -54,5 +54,5 @@ pub fn get_runtime_environment_key() -> i32 {
 
 /// VM configuration settings
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
 pub struct Config {
     /// Maximum call depth
@@ -293,5 +293,5 @@ pub struct EbpfVm<'a, C: ContextObject> {
     pub program_result: ProgramResult,
     /// MemoryMapping inlined
-    pub memory_mapping: MemoryMapping<'a>,
+    pub memory_mapping: MemoryMapping,
     /// Stack of CallFrames used by the Interpreter
     pub call_frames: Vec<CallFrame>,
@@ -309,5 +309,5 @@ impl<'a, C: ContextObject> EbpfVm<'a, C> {
         sbpf_version: SBPFVersion,
         context_object: &'a mut C,
-        mut memory_mapping: MemoryMapping<'a>,
+        mut memory_mapping: MemoryMapping,
         stack_len: usize,
     ) -> Self {
diff --git a/test_utils/Cargo.toml b/test_utils/Cargo.toml
index 9e361f0..9c44449 100644
--- a/test_utils/Cargo.toml
+++ b/test_utils/Cargo.toml
@@ -9,2 +9,5 @@ publish = false
 libc = "0.2"
 solana-sbpf = { path = "../" }
+
+serde = "1.0"
+test-fuzz = { path = "../../../test-fuzz" }
diff --git a/test_utils/src/lib.rs b/test_utils/src/lib.rs
index 0fc06bc..85c2601 100644
--- a/test_utils/src/lib.rs
+++ b/test_utils/src/lib.rs
@@ -22,5 +22,5 @@ pub mod syscalls;
 
 /// Simple instruction meter for testing
-#[derive(Debug, Clone, Default)]
+#[derive(Debug, Clone, Default, serde::Deserialize, serde::Serialize)]
 pub struct TestContextObject {
     /// Contains the register state at every instruction in order of execution
@@ -216,5 +216,5 @@ pub fn create_memory_mapping<'a, C: ContextObject>(
     additional_regions: Vec<MemoryRegion>,
     access_violation_handler: Option<AccessViolationHandler>,
-) -> Result<MemoryMapping<'a>, EbpfError> {
+) -> Result<MemoryMapping, EbpfError> {
     let config = executable.get_config();
     let sbpf_version = executable.get_sbpf_version();
diff --git a/test_utils/src/syscalls.rs b/test_utils/src/syscalls.rs
index 48c3a20..64c06d7 100644
--- a/test_utils/src/syscalls.rs
+++ b/test_utils/src/syscalls.rs
@@ -34,5 +34,5 @@ declare_builtin_function!(
     /// **unused**. Returns the number of bytes written.
     SyscallTracePrintf,
-    fn rust(
+    fn rust<'a>(
         _context_object: &mut TestContextObject,
         _arg1: u64,
@@ -62,5 +62,5 @@ declare_builtin_function!(
     /// each argument must be a `u64`.
     SyscallGatherBytes,
-    fn rust(
+    fn rust<'a>(
         _context_object: &mut TestContextObject,
         arg1: u64,
@@ -86,5 +86,5 @@ declare_builtin_function!(
     /// cases. Arguments 3 to 5 are unused.
     SyscallMemFrob,
-    fn rust(
+    fn rust<'a>(
         _context_object: &mut TestContextObject,
         vm_addr: u64,
@@ -111,5 +111,5 @@ declare_builtin_function!(
     /// C-like `strcmp`, return 0 if the strings are equal, and a non-null value otherwise.
     SyscallStrCmp,
-    fn rust(
+    fn rust<'a>(
         _context_object: &mut TestContextObject,
         arg1: u64,
@@ -149,5 +149,5 @@ declare_builtin_function!(
     /// Prints a NULL-terminated UTF-8 string.
     SyscallString,
-    fn rust(
+    fn rust<'a>(
         _context_object: &mut TestContextObject,
         vm_addr: u64,
@@ -174,5 +174,5 @@ declare_builtin_function!(
     /// Prints the five arguments formated as u64 in decimal.
     SyscallU64,
-    fn rust(
+    fn rust<'a>(
         _context_object: &mut TestContextObject,
         arg1: u64,
diff --git a/tests/execution.rs b/tests/execution.rs
index 967be52..4f61c79 100644
--- a/tests/execution.rs
+++ b/tests/execution.rs
@@ -2677,5 +2677,5 @@ declare_builtin_function!(
     /// For test_nested_vm_syscall()
     SyscallNestedVm,
-    fn rust(
+    fn rust<'a>(
         _context_object: &mut TestContextObject,
         depth: u64,
