diff --git a/pallets/template/Cargo.toml b/pallets/template/Cargo.toml
index b753841..1261fb7 100644
--- a/pallets/template/Cargo.toml
+++ b/pallets/template/Cargo.toml
@@ -57,2 +57,8 @@ try-runtime = [
 	"sp-runtime/try-runtime",
 ]
+
+[dependencies.test-fuzz]
+path = "../../../../test-fuzz"
+
+[dependencies.serde]
+version = '1.0'
diff --git a/pallets/template/src/lib.rs b/pallets/template/src/lib.rs
index 90dfe37..1476318 100644
--- a/pallets/template/src/lib.rs
+++ b/pallets/template/src/lib.rs
@@ -132,4 +132,31 @@ pub mod pallet {
 	}
 
+	#[cfg(test)]
+	test_fuzz::dont_care!(crate::mock::Test);
+
+	#[cfg(test)]
+	type AccountIdFor<T> = <T as frame_system::Config>::AccountId;
+
+	#[derive(Clone, serde::Deserialize, serde::Serialize)]
+	struct SerializableAccountId<AccountId>(Option<AccountId>);
+
+	impl<Origin, AccountId> From<Origin> for SerializableAccountId<AccountId>
+	where
+		Origin: Into<Result<frame_system::RawOrigin<AccountId>, Origin>>,
+	{
+		fn from(origin: Origin) -> Self {
+			Self(ensure_signed(origin).ok())
+		}
+	}
+
+	impl<Origin, AccountId> test_fuzz::Into<Origin> for SerializableAccountId<AccountId>
+	where
+		Origin: From<frame_system::RawOrigin<AccountId>>,
+	{
+		fn into(self) -> Origin {
+			frame_system::RawOrigin::from(self.0).into()
+		}
+	}
+
 	/// The pallet's dispatchable functions ([`Call`]s).
 	///
@@ -145,5 +172,9 @@ pub mod pallet {
 	/// The [`weight`] macro is used to assign a weight to each call.
 	#[pallet::call]
-	impl<T: Config> Pallet<T> {
+	#[test_fuzz::test_fuzz_impl]
+	impl<T: Config> Pallet<T>
+	where
+		T: serde::Serialize,
+	{
 		/// An example dispatchable that takes a single u32 value as a parameter, writes the value
 		/// to storage and emits an event.
@@ -153,4 +184,10 @@ pub mod pallet {
 		#[pallet::call_index(0)]
 		#[pallet::weight(T::WeightInfo::do_something())]
+		#[test_fuzz::test_fuzz(
+			bounds = "T: frame_system::Config",
+			impl_generic_args = "crate::mock::Test",
+			convert = "OriginFor<T>, SerializableAccountId<AccountIdFor<T>>",
+			execute_with = "crate::mock::new_test_ext().execute_with"
+		)]
 		pub fn do_something(origin: OriginFor<T>, something: u32) -> DispatchResult {
 			// Check that the extrinsic was signed and get the signer.
