diff --git a/substrate/client/transaction-pool/Cargo.toml b/substrate/client/transaction-pool/Cargo.toml
index 4e4ec9e..9e1b296 100644
--- a/substrate/client/transaction-pool/Cargo.toml
+++ b/substrate/client/transaction-pool/Cargo.toml
@@ -28,10 +28,10 @@ indexmap = { workspace = true }
 itertools = { workspace = true }
 linked-hash-map = { workspace = true }
-parking_lot = { workspace = true, default-features = true }
+parking_lot = { workspace = true, features = ["serde"] }
 prometheus-endpoint = { workspace = true, default-features = true }
 sc-client-api = { workspace = true, default-features = true }
 sc-transaction-pool-api = { workspace = true, default-features = true }
 sc-utils = { workspace = true, default-features = true }
-serde = { features = ["derive"], workspace = true, default-features = true }
+serde = { features = ["derive", "rc"], workspace = true, default-features = true }
 sp-api = { workspace = true, default-features = true }
 sp-blockchain = { workspace = true, default-features = true }
@@ -46,4 +46,6 @@ tokio-stream = { workspace = true }
 tracing = { workspace = true, default-features = true }
 
+test-fuzz = { path = "../../../../../test-fuzz" }
+
 [dev-dependencies]
 anyhow = { workspace = true }
diff --git a/substrate/client/transaction-pool/src/graph/base_pool.rs b/substrate/client/transaction-pool/src/graph/base_pool.rs
index 27d7878..7769dbe 100644
--- a/substrate/client/transaction-pool/src/graph/base_pool.rs
+++ b/substrate/client/transaction-pool/src/graph/base_pool.rs
@@ -85,5 +85,5 @@ pub struct PruneStatus<Hash, Ex> {
 
 /// A transaction source that includes a timestamp indicating when the transaction was submitted.
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
 pub struct TimedTransactionSource {
 	/// The original source of the transaction.
@@ -91,4 +91,5 @@ pub struct TimedTransactionSource {
 
 	/// The time at which the transaction was submitted.
+	#[serde(skip)]
 	pub timestamp: Option<Instant>,
 }
@@ -123,5 +124,5 @@ impl TimedTransactionSource {
 
 /// Immutable transaction
-#[derive(PartialEq, Eq, Clone)]
+#[derive(PartialEq, Eq, Clone, serde::Deserialize, serde::Serialize)]
 pub struct Transaction<Hash, Extrinsic> {
 	/// Raw extrinsic representing that transaction.
@@ -259,5 +260,7 @@ pub struct BasePool<Hash: hash::Hash + Eq, Ex> {
 }
 
-impl<Hash: hash::Hash + Member + Serialize, Ex: std::fmt::Debug> Default for BasePool<Hash, Ex> {
+impl<Hash: hash::Hash + Member + Serialize, Ex: Clone + std::fmt::Debug + Serialize> Default
+	for BasePool<Hash, Ex>
+{
 	fn default() -> Self {
 		Self::new(false)
@@ -265,5 +268,7 @@ impl<Hash: hash::Hash + Member + Serialize, Ex: std::fmt::Debug> Default for Bas
 }
 
-impl<Hash: hash::Hash + Member + Serialize, Ex: std::fmt::Debug> BasePool<Hash, Ex> {
+impl<Hash: hash::Hash + Member + Serialize, Ex: Clone + std::fmt::Debug + Serialize>
+	BasePool<Hash, Ex>
+{
 	/// Create new pool given reject_future_transactions flag.
 	pub fn new(reject_future_transactions: bool) -> Self {
diff --git a/substrate/client/transaction-pool/src/graph/future.rs b/substrate/client/transaction-pool/src/graph/future.rs
index 848893b..8742168 100644
--- a/substrate/client/transaction-pool/src/graph/future.rs
+++ b/substrate/client/transaction-pool/src/graph/future.rs
@@ -30,4 +30,5 @@ use super::base_pool::Transaction;
 use crate::{common::tracing_log_xt::log_xt_trace, LOG_TARGET};
 
+#[derive(serde::Deserialize, serde::Serialize)]
 /// Transaction with partially satisfied dependencies.
 pub struct WaitingTransaction<Hash, Ex> {
@@ -37,4 +38,5 @@ pub struct WaitingTransaction<Hash, Ex> {
 	pub missing_tags: HashSet<Tag>,
 	/// Time of import to the Future Queue.
+	#[serde(skip, default = "Instant::now")]
 	pub imported_at: Instant,
 }
diff --git a/substrate/client/transaction-pool/src/graph/ready.rs b/substrate/client/transaction-pool/src/graph/ready.rs
index 88986ba..0ac9f7e 100644
--- a/substrate/client/transaction-pool/src/graph/ready.rs
+++ b/substrate/client/transaction-pool/src/graph/ready.rs
@@ -39,5 +39,5 @@ use super::{
 ///
 /// Should be cheap to clone.
-#[derive(Debug)]
+#[derive(Debug, serde::Deserialize, serde::Serialize)]
 pub struct TransactionRef<Hash, Ex> {
 	/// The actual transaction data.
@@ -76,5 +76,5 @@ impl<Hash, Ex> PartialEq for TransactionRef<Hash, Ex> {
 impl<Hash, Ex> Eq for TransactionRef<Hash, Ex> {}
 
-#[derive(Debug)]
+#[derive(Debug, serde::Deserialize, serde::Serialize)]
 pub struct ReadyTx<Hash, Ex> {
 	/// A reference to a transaction
@@ -107,5 +107,5 @@ qed
 
 /// Validated transactions that are block ready with all their dependencies met.
-#[derive(Clone, Debug)]
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize)]
 pub struct ReadyTransactions<Hash: hash::Hash + Eq, Ex> {
 	/// Next free insertion id (used to indicate when a transaction was inserted into the pool).
@@ -138,5 +138,10 @@ impl<Hash: hash::Hash + Eq, Ex> Default for ReadyTransactions<Hash, Ex> {
 }
 
-impl<Hash: hash::Hash + Member + Serialize, Ex> ReadyTransactions<Hash, Ex> {
+#[test_fuzz::test_fuzz_impl]
+impl<
+		Hash: std::fmt::Debug + hash::Hash + Member + Serialize,
+		Ex: Clone + std::fmt::Debug + Serialize,
+	> ReadyTransactions<Hash, Ex>
+{
 	/// Borrows a map of tags that are provided by transactions in this queue.
 	pub fn provided_tags(&self) -> &HashMap<Tag, Hash> {
@@ -176,8 +181,12 @@ impl<Hash: hash::Hash + Member + Serialize, Ex> ReadyTransactions<Hash, Ex> {
 	/// that are in this queue.
 	/// Returns transactions that were replaced by the one imported.
+	#[test_fuzz::test_fuzz(impl_generic_args = "u64, Vec<u8>", bounds = "Hash: Eq + hash::Hash")]
 	pub fn import(
 		&mut self,
 		tx: WaitingTransaction<Hash, Ex>,
 	) -> error::Result<Vec<Arc<Transaction<Hash, Ex>>>> {
+		if !tx.is_ready() {
+			return Ok(Vec::default());
+		}
 		assert!(
 			tx.is_ready(),
@@ -185,4 +194,7 @@ impl<Hash: hash::Hash + Member + Serialize, Ex> ReadyTransactions<Hash, Ex> {
 			tx.missing_tags
 		);
+		if self.ready.read().contains_key(&tx.transaction.hash) {
+			return Ok(Vec::default());
+		}
 		assert!(
 			!self.ready.read().contains_key(&tx.transaction.hash),
@@ -204,4 +216,7 @@ impl<Hash: hash::Hash + Member + Serialize, Ex> ReadyTransactions<Hash, Ex> {
 			// Check if the transaction that satisfies the tag is still in the queue.
 			if let Some(other) = self.provided_tags.get(tag) {
+				if ready.get_mut(other).is_none() {
+					return Ok(Vec::default());
+				}
 				let tx = ready.get_mut(other).expect(HASH_READY);
 				tx.unlocks.push(hash.clone());
@@ -604,5 +619,8 @@ mod tests {
 	}
 
-	fn import<H: hash::Hash + Eq + Member + Serialize, Ex>(
+	fn import<
+		H: std::fmt::Debug + hash::Hash + Eq + Member + Serialize,
+		Ex: Clone + std::fmt::Debug + Serialize,
+	>(
 		ready: &mut ReadyTransactions<H, Ex>,
 		tx: Transaction<H, Ex>,
diff --git a/substrate/client/transaction-pool/src/graph/tracked_map.rs b/substrate/client/transaction-pool/src/graph/tracked_map.rs
index e9bdd1f..796d6f5 100644
--- a/substrate/client/transaction-pool/src/graph/tracked_map.rs
+++ b/substrate/client/transaction-pool/src/graph/tracked_map.rs
@@ -34,6 +34,9 @@ pub trait Size {
 ///
 /// Size reported might be slightly off and only approximately true.
-#[derive(Debug)]
-pub struct TrackedMap<K, V> {
+#[derive(Debug, serde::Deserialize, serde::Serialize)]
+pub struct TrackedMap<K, V>
+where
+	K: Eq + std::hash::Hash,
+{
 	index: Arc<RwLock<HashMap<K, V>>>,
 	bytes: AtomicIsize,
@@ -41,5 +44,8 @@ pub struct TrackedMap<K, V> {
 }
 
-impl<K, V> Default for TrackedMap<K, V> {
+impl<K, V> Default for TrackedMap<K, V>
+where
+	K: Eq + std::hash::Hash,
+{
 	fn default() -> Self {
 		Self { index: Arc::new(HashMap::default().into()), bytes: 0.into(), length: 0.into() }
@@ -49,5 +55,5 @@ impl<K, V> Default for TrackedMap<K, V> {
 impl<K, V> Clone for TrackedMap<K, V>
 where
-	K: Clone,
+	K: Clone + Eq + std::hash::Hash,
 	V: Clone,
 {
@@ -61,5 +67,8 @@ where
 }
 
-impl<K, V> TrackedMap<K, V> {
+impl<K, V> TrackedMap<K, V>
+where
+	K: Eq + std::hash::Hash,
+{
 	/// Current tracked length of the content.
 	pub fn len(&self) -> usize {
@@ -87,5 +96,8 @@ impl<K, V> TrackedMap<K, V> {
 }
 
-impl<K: Clone, V: Clone> TrackedMap<K, V> {
+impl<K: Clone, V: Clone> TrackedMap<K, V>
+where
+	K: Eq + std::hash::Hash,
+{
 	/// Clone the inner map.
 	pub fn clone_map(&self) -> HashMap<K, V> {
diff --git a/substrate/primitives/runtime/src/transaction_validity.rs b/substrate/primitives/runtime/src/transaction_validity.rs
index 4fe69d6..867ab81 100644
--- a/substrate/primitives/runtime/src/transaction_validity.rs
+++ b/substrate/primitives/runtime/src/transaction_validity.rs
@@ -242,4 +242,5 @@ impl From<UnknownTransaction> for TransactionValidity {
 /// by our local node (for instance off-chain workers).
 #[derive(Copy, Clone, PartialEq, Eq, Encode, Decode, RuntimeDebug, TypeInfo, Hash)]
+#[cfg_attr(feature = "std", derive(serde::Deserialize, serde::Serialize))]
 pub enum TransactionSource {
 	/// Transaction is already included in block.
