diff --git a/client/transaction-pool/Cargo.toml b/client/transaction-pool/Cargo.toml
index 0bdfb62..12745e7 100644
--- a/client/transaction-pool/Cargo.toml
+++ b/client/transaction-pool/Cargo.toml
@@ -21,6 +21,6 @@ linked-hash-map = "0.5.4"
 log = "0.4.17"
 parity-util-mem = { version = "0.12.0", default-features = false, features = ["primitive-types"] }
-parking_lot = "0.12.1"
-serde = { version = "1.0.136", features = ["derive"] }
+parking_lot = { version = "0.12.0", features = ["serde"] }
+serde = { version = "1.0.136", features = ["derive", "rc"] }
 thiserror = "1.0.30"
 prometheus-endpoint = { package = "substrate-prometheus-endpoint", version = "0.10.0-dev", path = "../../utils/prometheus" }
@@ -34,4 +34,5 @@ sp-runtime = { version = "6.0.0", path = "../../primitives/runtime" }
 sp-tracing = { version = "5.0.0", path = "../../primitives/tracing" }
 sp-transaction-pool = { version = "4.0.0-dev", path = "../../primitives/transaction-pool" }
+test-fuzz = { path = "../../../../test-fuzz" }
 
 [dev-dependencies]
diff --git a/client/transaction-pool/src/graph/base_pool.rs b/client/transaction-pool/src/graph/base_pool.rs
index 8e04227..af1b2c5 100644
--- a/client/transaction-pool/src/graph/base_pool.rs
+++ b/client/transaction-pool/src/graph/base_pool.rs
@@ -85,5 +85,5 @@ pub struct PruneStatus<Hash, Ex> {
 /// Immutable transaction
 #[cfg_attr(test, derive(Clone))]
-#[derive(PartialEq, Eq, parity_util_mem::MallocSizeOf)]
+#[derive(PartialEq, Eq, parity_util_mem::MallocSizeOf, serde::Deserialize, serde::Serialize)]
 pub struct Transaction<Hash, Extrinsic> {
 	/// Raw extrinsic representing that transaction.
@@ -221,5 +221,7 @@ pub struct BasePool<Hash: hash::Hash + Eq, Ex> {
 }
 
-impl<Hash: hash::Hash + Member + Serialize, Ex: std::fmt::Debug> Default for BasePool<Hash, Ex> {
+impl<Hash: hash::Hash + Member + Serialize, Ex: Clone + std::fmt::Debug + Serialize> Default
+	for BasePool<Hash, Ex>
+{
 	fn default() -> Self {
 		Self::new(false)
@@ -227,5 +229,7 @@ impl<Hash: hash::Hash + Member + Serialize, Ex: std::fmt::Debug> Default for Bas
 }
 
-impl<Hash: hash::Hash + Member + Serialize, Ex: std::fmt::Debug> BasePool<Hash, Ex> {
+impl<Hash: hash::Hash + Member + Serialize, Ex: Clone + std::fmt::Debug + Serialize>
+	BasePool<Hash, Ex>
+{
 	/// Create new pool given reject_future_transactions flag.
 	pub fn new(reject_future_transactions: bool) -> Self {
diff --git a/client/transaction-pool/src/graph/future.rs b/client/transaction-pool/src/graph/future.rs
index ae49e3f..4f7f4bf 100644
--- a/client/transaction-pool/src/graph/future.rs
+++ b/client/transaction-pool/src/graph/future.rs
@@ -29,5 +29,5 @@ use std::time::Instant;
 use super::base_pool::Transaction;
 
-#[derive(parity_util_mem::MallocSizeOf)]
+#[derive(parity_util_mem::MallocSizeOf, serde::Deserialize, serde::Serialize)]
 /// Transaction with partially satisfied dependencies.
 pub struct WaitingTransaction<Hash, Ex> {
@@ -37,4 +37,5 @@ pub struct WaitingTransaction<Hash, Ex> {
 	pub missing_tags: HashSet<Tag>,
 	/// Time of import to the Future Queue.
+	#[serde(skip, default = "Instant::now")]
 	pub imported_at: Instant,
 }
diff --git a/client/transaction-pool/src/graph/ready.rs b/client/transaction-pool/src/graph/ready.rs
index 220e69b..79bce8a 100644
--- a/client/transaction-pool/src/graph/ready.rs
+++ b/client/transaction-pool/src/graph/ready.rs
@@ -38,5 +38,5 @@ use super::{
 ///
 /// Should be cheap to clone.
-#[derive(Debug, parity_util_mem::MallocSizeOf)]
+#[derive(Debug, parity_util_mem::MallocSizeOf, serde::Deserialize, serde::Serialize)]
 pub struct TransactionRef<Hash, Ex> {
 	/// The actual transaction data.
@@ -75,5 +75,5 @@ impl<Hash, Ex> PartialEq for TransactionRef<Hash, Ex> {
 impl<Hash, Ex> Eq for TransactionRef<Hash, Ex> {}
 
-#[derive(Debug, parity_util_mem::MallocSizeOf)]
+#[derive(Debug, parity_util_mem::MallocSizeOf, serde::Deserialize, serde::Serialize)]
 pub struct ReadyTx<Hash, Ex> {
 	/// A reference to a transaction
@@ -106,5 +106,5 @@ qed
 
 /// Validated transactions that are block ready with all their dependencies met.
-#[derive(Debug, parity_util_mem::MallocSizeOf)]
+#[derive(Clone, Debug, parity_util_mem::MallocSizeOf, serde::Deserialize, serde::Serialize)]
 pub struct ReadyTransactions<Hash: hash::Hash + Eq, Ex> {
 	/// Next free insertion id (used to indicate when a transaction was inserted into the pool).
@@ -137,5 +137,10 @@ impl<Hash: hash::Hash + Eq, Ex> Default for ReadyTransactions<Hash, Ex> {
 }
 
-impl<Hash: hash::Hash + Member + Serialize, Ex> ReadyTransactions<Hash, Ex> {
+#[test_fuzz::test_fuzz_impl]
+impl<
+		Hash: std::fmt::Debug + hash::Hash + Member + Serialize,
+		Ex: Clone + std::fmt::Debug + Serialize,
+	> ReadyTransactions<Hash, Ex>
+{
 	/// Borrows a map of tags that are provided by transactions in this queue.
 	pub fn provided_tags(&self) -> &HashMap<Tag, Hash> {
@@ -175,8 +180,15 @@ impl<Hash: hash::Hash + Member + Serialize, Ex> ReadyTransactions<Hash, Ex> {
 	/// that are in this queue.
 	/// Returns transactions that were replaced by the one imported.
+	#[test_fuzz::test_fuzz(
+		concretize_impl = "sp_core::H256, substrate_test_runtime::Extrinsic",
+		bounds = "Hash: Eq + hash::Hash"
+	)]
 	pub fn import(
 		&mut self,
 		tx: WaitingTransaction<Hash, Ex>,
 	) -> error::Result<Vec<Arc<Transaction<Hash, Ex>>>> {
+		if !tx.is_ready() {
+			return Ok(Vec::default());
+		}
 		assert!(
 			tx.is_ready(),
@@ -184,4 +196,7 @@ impl<Hash: hash::Hash + Member + Serialize, Ex> ReadyTransactions<Hash, Ex> {
 			tx.missing_tags
 		);
+		if self.ready.read().contains_key(&tx.transaction.hash) {
+			return Ok(Vec::default());
+		}
 		assert!(
 			!self.ready.read().contains_key(&tx.transaction.hash),
@@ -203,4 +218,7 @@ impl<Hash: hash::Hash + Member + Serialize, Ex> ReadyTransactions<Hash, Ex> {
 			// Check if the transaction that satisfies the tag is still in the queue.
 			if let Some(other) = self.provided_tags.get(tag) {
+				if ready.get_mut(other).is_none() {
+					return Ok(Vec::default());
+				}
 				let tx = ready.get_mut(other).expect(HASH_READY);
 				tx.unlocks.push(hash.clone());
@@ -606,5 +624,8 @@ mod tests {
 	}
 
-	fn import<H: hash::Hash + Eq + Member + Serialize, Ex>(
+	fn import<
+		H: std::fmt::Debug + hash::Hash + Eq + Member + Serialize,
+		Ex: Clone + std::fmt::Debug + Serialize,
+	>(
 		ready: &mut ReadyTransactions<H, Ex>,
 		tx: Transaction<H, Ex>,
diff --git a/client/transaction-pool/src/graph/tracked_map.rs b/client/transaction-pool/src/graph/tracked_map.rs
index 32d04b0..3691ef0 100644
--- a/client/transaction-pool/src/graph/tracked_map.rs
+++ b/client/transaction-pool/src/graph/tracked_map.rs
@@ -34,6 +34,9 @@ pub trait Size {
 ///
 /// Size reported might be slightly off and only approximately true.
-#[derive(Debug, parity_util_mem::MallocSizeOf)]
-pub struct TrackedMap<K, V> {
+#[derive(Debug, parity_util_mem::MallocSizeOf, serde::Deserialize, serde::Serialize)]
+pub struct TrackedMap<K, V>
+where
+	K: Eq + std::hash::Hash,
+{
 	index: Arc<RwLock<HashMap<K, V>>>,
 	bytes: AtomicIsize,
@@ -41,5 +44,21 @@ pub struct TrackedMap<K, V> {
 }
 
-impl<K, V> Default for TrackedMap<K, V> {
+impl<K: Clone, V: Clone> Clone for TrackedMap<K, V>
+where
+	K: Eq + std::hash::Hash,
+{
+	fn clone(&self) -> Self {
+		Self {
+			index: Arc::new(self.clone_map().into()),
+			bytes: self.bytes.load(AtomicOrdering::Relaxed).into(),
+			length: self.length.load(AtomicOrdering::Relaxed).into(),
+		}
+	}
+}
+
+impl<K, V> Default for TrackedMap<K, V>
+where
+	K: Eq + std::hash::Hash,
+{
 	fn default() -> Self {
 		Self { index: Arc::new(HashMap::default().into()), bytes: 0.into(), length: 0.into() }
@@ -47,5 +66,8 @@ impl<K, V> Default for TrackedMap<K, V> {
 }
 
-impl<K, V> TrackedMap<K, V> {
+impl<K, V> TrackedMap<K, V>
+where
+	K: Eq + std::hash::Hash,
+{
 	/// Current tracked length of the content.
 	pub fn len(&self) -> usize {
@@ -73,5 +95,8 @@ impl<K, V> TrackedMap<K, V> {
 }
 
-impl<K: Clone, V: Clone> TrackedMap<K, V> {
+impl<K: Clone, V: Clone> TrackedMap<K, V>
+where
+	K: Eq + std::hash::Hash,
+{
 	/// Clone the inner map.
 	pub fn clone_map(&self) -> HashMap<K, V> {
diff --git a/primitives/runtime/src/transaction_validity.rs b/primitives/runtime/src/transaction_validity.rs
index 4646808..e8e6cc0 100644
--- a/primitives/runtime/src/transaction_validity.rs
+++ b/primitives/runtime/src/transaction_validity.rs
@@ -228,5 +228,8 @@ impl From<UnknownTransaction> for TransactionValidity {
 /// by our local node (for instance off-chain workers).
 #[derive(Copy, Clone, PartialEq, Eq, Encode, Decode, RuntimeDebug)]
-#[cfg_attr(feature = "std", derive(parity_util_mem::MallocSizeOf))]
+#[cfg_attr(
+	feature = "std",
+	derive(parity_util_mem::MallocSizeOf, serde::Deserialize, serde::Serialize)
+)]
 pub enum TransactionSource {
 	/// Transaction is already included in block.
diff --git a/test-utils/runtime/src/lib.rs b/test-utils/runtime/src/lib.rs
index 8bda4ea..64f296f 100644
--- a/test-utils/runtime/src/lib.rs
+++ b/test-utils/runtime/src/lib.rs
@@ -186,5 +186,5 @@ impl<'a> serde::Deserialize<'a> for Extrinsic {
 		D: serde::Deserializer<'a>,
 	{
-		let r = sp_core::bytes::deserialize(de)?;
+		let r = <Vec<u8>>::deserialize(de)?;
 		Decode::decode(&mut &r[..])
 			.map_err(|e| serde::de::Error::custom(format!("Decode error: {}", e)))
